### Курсовой проект. Дисперсионный анализ.

Датафрейм взят из сайта https://www.kaggle.com/competitions/house-prices-advanced-regression-techniques

In [30]:
import pandas as pd
import numpy as np
from scipy import stats

In [31]:
df = pd.read_csv('./df_course_project.csv')

In [32]:
df.sample(5)

MSZoning  SalePrice
1203       RL     213000
187        RL     135000
616        RL     183200
1002       RL     232000
886        RL     145000

**MSZoning**: определяет общую классификацию зонирования продажи.

RL - Жилая Низкая плотность

RM - Жилая Средняя плотность

C (all) - Коммерческая

FV - Плавучая деревня Жилая

RH - Жилая Высокая плотность

**SalePrice**: Цена продаж.

In [33]:
df['MSZoning'].unique()

array(['RL', 'RM', 'C (all)', 'FV', 'RH'], dtype=object)

In [34]:
df['MSZoning'].value_counts()

RL         1151
RM          218
FV           65
RH           16
C (all)      10
Name: MSZoning, dtype: int64

In [35]:
len(df['MSZoning'].unique())

5

In [36]:
len(df['MSZoning'])

1460

In [37]:
f_a = df.loc[df['MSZoning'] == 'RL']
f_b = df.loc[df['MSZoning'] == 'RM']
f_c = df.loc[df['MSZoning'] == 'FV']
f_d = df.loc[df['MSZoning'] == 'RH']
f_e = df.loc[df['MSZoning'] == 'C (all)']

In [38]:
k = 5
n = 1460

In [39]:
# Среднее по каждой выборке
a_mean = f_a['SalePrice'].mean()
b_mean = f_b['SalePrice'].mean()
c_mean = f_c['SalePrice'].mean()
d_mean = f_d['SalePrice'].mean()
e_mean = f_e['SalePrice'].mean()

In [40]:
# Среднее по всей выборке
df_mean = df['SalePrice'].mean()

In [41]:
# Сумма квадратов отклонений средних групповых значений от общего среднего
ss_f = (a_mean - df_mean)**2 * len(f_a) + \
       (b_mean - df_mean)**2 * len(f_b) + \
       (c_mean - df_mean)**2 * len(f_c) + \
       (d_mean - df_mean)**2 * len(f_d) + \
       (e_mean - df_mean)**2 * len(f_e)
ss_f

990400028351.7827

In [42]:
# Остаточная сумма квадратов отклонений

In [43]:
ss_r_a = 0
for i in (f_a['SalePrice']):
    ss_r_a += (i - a_mean)**2
ss_r_a

7501682173621.974

In [44]:
ss_r_b = 0
for i in (f_b['SalePrice']):
    ss_r_b += (i - b_mean)**2
ss_r_b

510894901372.7202

In [45]:
ss_r_c = 0
for i in (f_c['SalePrice']):
    ss_r_c += (i - c_mean)**2
ss_r_c

175525216323.7538

In [46]:
ss_r_d = 0
for i in (f_d['SalePrice']):
    ss_r_d += (i - d_mean)**2
ss_r_d

19132473833.75

In [47]:
ss_r_e = 0
for i in (f_e['SalePrice']):
    ss_r_e += (i - e_mean)**2
ss_r_e

10276541106.0

In [19]:
ss_r = ss_r_a + ss_r_b + ss_r_c + ss_r_d + ss_r_e
ss_r

8217511306258.197

In [20]:
ss = ss_f + ss_r
ss

9207911334609.98

In [21]:
d_f1 = k - 1
d_f1

4

In [22]:
d_f2 = n - k
d_f2

1455

In [23]:
sdt_f = ss_f / d_f1
sdt_f

247600007087.94568

In [24]:
sdt_r = ss_r / d_f2
sdt_r

5647774093.648246

In [25]:
F_n = sdt_f / sdt_r
F_n

43.84028167245719

In [26]:
stats.f.ppf(0.95, d_f1, d_f2)

2.3780442396567194

Расчетное значение критерия Фишера = 43,84.

Табличное значение критерия Фишера = 2,38.

Следовательно, принимаем альтернативную гипотезу. Классификация зонирования продаж **оказывает** статистически значимый эффект на стоимость помещения.

Проверка с помощью функции.

In [29]:
stats.f_oneway(f_a['SalePrice'], f_b['SalePrice'], f_c['SalePrice'], \
              f_d['SalePrice'], f_e['SalePrice'])

F_onewayResult(statistic=43.84028167245718, pvalue=8.817633866272648e-35)

### Задача 1.

Провести дисперсионный анализ для определения того, есть ли различия среднего роста среди взрослых футболистов, хоккеистов и штангистов. Даны значения роста в трех группах случайно выбранных спортсменов: Футболисты: 173, 175, 180, 178, 177, 185, 183, 182. Хоккеисты: 177, 179, 180, 188, 177, 172, 171, 184, 180. Штангисты: 172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170. alpha = 0.05.

a - футболисты.

b - хоккеисты.

c - штангисты.

In [30]:
a = np.array([173, 175, 180, 178, 177, 185, 183, 182])
b = np.array([177, 179, 180, 188, 177, 172, 171, 184, 180])
c = np.array([172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170])

In [31]:
stats.f_oneway(a, b, c)

F_onewayResult(statistic=5.500053450812596, pvalue=0.010482206918698694)

In [32]:
k = 3
n = len(a) + len(b) + len(c)

In [35]:
# Среднее по каждой выборке
a_mean = a.mean()
b_mean = b.mean()
c_mean = c.mean()

In [38]:
# Среднее по всей выборке
y = np.concatenate ((a, b, c), axis=0)
y_mean = y.mean()
y_mean

176.46428571428572

In [39]:
# Сумма квадратов отклонений средних групповых значений от общего среднего
ss_f = (a_mean - y_mean)**2 * len(a) + \
       (b_mean - y_mean)**2 * len(b) + \
       (c_mean - y_mean)**2 * len(c)
ss_f

253.9074675324678

In [41]:
# Остаточная сумма квадратов отклонений

In [43]:
ss_r_a = 0
for i in (a):
    ss_r_a += (i - a_mean)**2
ss_r_a

118.875

In [44]:
ss_r_b = 0
for i in (b):
    ss_r_b += (i - b_mean)**2
ss_r_b

227.99999999999997

In [45]:
ss_r_c = 0
for i in (c):
    ss_r_c += (i - c_mean)**2
ss_r_c

230.18181818181816

In [46]:
ss_r = ss_r_a + ss_r_b + ss_r_c
ss_r

577.0568181818181

In [47]:
ss = ss_f + ss_r
ss

830.9642857142859

In [48]:
d_f1 = k - 1
d_f1

2

In [49]:
d_f2 = n - k
d_f2

25

In [50]:
sdt_f = ss_f / d_f1
sdt_f

126.9537337662339

In [51]:
sdt_r = ss_r / d_f2
sdt_r

23.082272727272724

In [52]:
F_n = sdt_f / sdt_r
F_n

5.500053450812599

In [54]:
stats.f.ppf(0.95, d_f1, d_f2)

3.3851899614491687

In [53]:
stats.f_oneway(a, b, c)

F_onewayResult(statistic=5.500053450812596, pvalue=0.010482206918698694)

Расчетное значение критерия Фишера = 5.5.

Табличное значение критерия Фишера = 3.38.

Следовательно, принимаем альтернативную гипотезу. То есть **существуют** статистически значимые различия среднего роста среди взрослых футболистов, хоккеистов и штангистов.